content/tokio/topics/tracing.md

```diff
@@ -0,0 +1,249 @@
+information about *temporality* and *causality* — unlike a log message, a
+[`Span`] in `tracing` has a beginning and end time, may be entered and exited
+by the flow of execution, and may exist within a nested tree of similar spans.
+For representing things that occur at a single moment in time, `tracing`
+provides the complementary concept of *events*. Both [`Span`]s and [`Event`]s
+are *structured*, with the ability to record typed data as well as textual
+messages.
+
+[`Span`]: https://docs.rs/tracing/latest/tracing/#spans
+[`Event`]: https://docs.rs/tracing/latest/tracing/#events
+
+You can use `tracing` to:
+- emit distributed traces to an [OpenTelemetry] collector
+- debug your application with [Tokio Console]
+- log to [`stdout`], [a log file] or [`journald`]
+- [profile] where your application is spending time
+
+[`tracing`]: https://docs.rs/tracing
+[`tracing-subscriber`]: https://docs.rs/tracing-subscriber
+[OpenTelemetry]: https://docs.rs/tracing-opentelemetry
+[Tokio Console]: https://docs.rs/console-subscriber
+[`stdout`]: https://docs.rs/tracing-subscriber/latest/tracing_subscriber/fmt/index.html
+[a log file]: https://docs.rs/tracing-appender/latest/tracing_appender/
+[`journald`]: https://docs.rs/tracing-journald/latest/tracing_journald/
+[profile]: https://docs.rs/tracing-timing/latest/tracing_timing/
+
+# Setup
+
+To begin, add [`tracing`] and [`tracing-subscriber`] as dependencies:
+
+```toml
+[dependencies]
+tracing = "0.1"
+tracing-subscriber = "0.3"
+```
+
+The [`tracing`] crate provides the API we will use to emit traces. The
+[`tracing-subscriber`] crate provides some basic utilities for forwarding those
+traces to external listeners (e.g., `stdout`).
+
+# Subscribing to Traces
+
+If you are authoring an executable (as opposed to a library), you will need to
+register a tracing *subscriber*. Subscribers are types that process traces
+emitted by your application and its dependencies, and and can perform tasks
+such as computing metrics, monitoring for errors, and re-emitting traces to the
+outside world (e.g., `journald`, `stdout`, or an `open-telemetry` daemon).
+
+In most circumstances, you should register your tracing subscriber as early as
+possible in your `main` function. For instance, the [`FmtSubscriber`] type
+provided by [`tracing-subscriber`] prints formatted traces and events to
+`stdout`, can be registered like so:
+
+```rust
+# mod mini_redis {
+#       pub type Error = Box<dyn std::error::Error + Send + Sync>;
+#       pub type Result<T> = std::result::Result<T, Error>;
+# }
+#[tokio::main]
+pub async fn main() -> mini_redis::Result<()> {
+    // construct a subscriber that prints formatted traces to stdout
+    let subscriber = tracing_subscriber::FmtSubscriber::new();
+    // use that subscriber to process traces emitted after this point
+    tracing::subscriber::set_global_default(subscriber)?;
+# /*
+    ...
+# */ Ok(())
+}
+```
```

Comment:
The hashed sections here and at the end of the block are kind of throwing me off (partially because my brain is trying to understand what the hashes indicate in Rust (nothing right? I think you were intending to comment-out that code in a way) and also because the line right after is the annotation which does use a hash..

Do we really need the `mod mini_redis` section?

I'd maybe just write it as

```rust
#[tokio::main]
pub async fn main() -> mini_redis::Result<()> {
    // construct a subscriber that prints formatted traces to stdout
    let subscriber = tracing_subscriber::FmtSubscriber::new();
    // use that subscriber to process traces emitted after this point
    tracing::subscriber::set_global_default(subscriber)?;

    ... // the rest of main

    Ok(())
}
```

Reply:
Lines beginning with `#` are hidden from the rendered output, but processed when the code blocks are run as doc tests.

Reply:
Oh I see, so the website won't show them? Gotcha -- I was just looking at the file directly as GitHub renders it. You can ignore my comments related to this then